lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -3072,11 +3087,64 @@ ConformanceChecker::inferTypeWitnessesViaValueWitnesses(
           if (!dmt->getBase()->isEqual(Conformance->getType()))
             return false;
           
+          // If this associated type is same-typed to another associated type
+          // on `Self`, then it may still be an interesting candidate if we find
+          // an answer for that other type.
+          if (witness->getDeclContext()->getAsProtocolExtensionContext()
```

Comment:
If you want to check that two specific associated types are equal, you can map both into context and compare archetypes for pointer equality.

To see if an associated type is equal to *some other* associated type, the only way that's slightly saner than yours is to get the canonical generic signature builder for the signature, and ask it for the equivalence class of the associated type. But I don't know if the right APIs are public or sane.

So what you have here might be your best bet.